<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="James Dalgleish" />

<meta name="date" content="2018-07-25" />

<title>Creating Linear Regression Matrices from Segment data</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Creating Linear Regression Matrices from Segment data</h1>
<h4 class="author">James Dalgleish</h4>
<h4 class="date">July 25, 2018</h4>



<p>From our previous work, we created a small input matrix, with segmented 1Mb regions as our row labels and with sample names from TARGET data as our column labels. We can read that in using the following code:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">library</span>(CNVScope)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>nbl_input_matrix[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>##                      PASEGA   PASRFS   PARHYL   PARVME PATGJU
## chr1_1_1000000         0.77 1.226667 0.990000 1.304444   1.09
## chr1_1000001_2000000   0.77 1.000000 1.053333 1.180000   1.09
## chr1_2000001_3000000   0.77 1.000000 0.850000 1.000000   1.09
## chr1_3000001_4000000   0.77 1.000000 0.850000 1.000000   1.09
## chr1_4000001_5000000   0.77 1.000000 0.850000 1.000000   1.09</code></pre>
<p>calcVecLMs() comes standard in the CNVScope package. It allows calculation of the matrix with parallel processing using mclapply, but larger matrices will require a bit more power, and thus we use slurm_apply, from the rslurm pacakge to distribute the work over multiple cores. Our particular establishment has a limit approximating 1000 jobs, so it’s best not to use more than that unless your cluster will support it. Conversely, you should use less if you can’t submit that many individual jobs in a job array in your cluster. In this particular example, I’ve removed rows where there is no segmentation data, across the board using colSums().</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="kw">library</span>(parallel)</span>
<span id="cb4-2"><a href="#cb4-2"></a>nbl_slurm_object_test_zero_removed&lt;-<span class="kw">calcVecLMs</span>(<span class="dt">bin_data =</span><span class="kw">as.data.frame</span>(<span class="kw">t</span>(nbl_input_matrix[<span class="kw">which</span>(<span class="kw">rowSds</span>(<span class="kw">as.matrix</span>(nbl_input_matrix))<span class="op">!=</span><span class="fl">0.0</span>),])),<span class="dt">use_slurm =</span> T,<span class="dt">n_nodes =</span> <span class="dv">975</span>,<span class="dt">memory_per_node =</span> <span class="st">&quot;32g&quot;</span>,<span class="dt">walltime =</span> <span class="st">&quot;04:00:00&quot;</span>,<span class="dt">cpus_on_each_node =</span> <span class="dv">2</span>,<span class="dt">job_finished =</span> F,<span class="dt">slurmjob =</span> <span class="ot">NULL</span>)</span></code></pre></div>
<p>Saving the slurm object is essential as it will be required when you retrieve your results.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="kw">saveRDS</span>(nbl_slurm_object_test_zero_removed,<span class="st">&quot;nbl_slurm_object_test_zero_removed.rds&quot;</span>)</span></code></pre></div>
<p>Retrieving the data is as simple as using rslurm::get_slurm_out() on the saved slurm object and coercing it into a matrix with the original number of columns. The slurm object must have been read with readRDS() previously or done in the same session. For the purposes of making this tutorial, we have chosen to work on a small version of the whole matrix to make 5MB CRAN documentation limits. Previous versions of the tutorial included the whole matrix, but we leave that to the user to construct at this point. For reproducibility, one can find the original full data matrix at <a href="https://github.com/jamesdalg/CNVScope_public_data" class="uri">https://github.com/jamesdalg/CNVScope_public_data</a>.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="kw">library</span>(matrixStats)</span>
<span id="cb6-2"><a href="#cb6-2"></a>nbl_result_matrix&lt;-<span class="kw">matrix</span>(<span class="kw">as.numeric</span>(<span class="kw">unlist</span>( <span class="kw">get_slurm_out</span>(nbl_slurm_object_test_zero_removed))),<span class="dt">ncol=</span><span class="kw">ncol</span>(<span class="kw">as.data.frame</span>(<span class="kw">t</span>(nbl_input_matrix[<span class="kw">which</span>(<span class="kw">rowSds</span>(<span class="kw">as.matrix</span>(nbl_input_matrix))<span class="op">!=</span><span class="fl">0.0</span>),])) ) )</span>
<span id="cb6-3"><a href="#cb6-3"></a></span>
<span id="cb6-4"><a href="#cb6-4"></a><span class="kw">saveRDS</span>(nbl_result_matrix,<span class="st">&quot;nbl_result_matrix_full.rds&quot;</span>)</span>
<span id="cb6-5"><a href="#cb6-5"></a><span class="kw">saveRDS</span>(nbl_result_matrix[<span class="dv">1</span><span class="op">:</span><span class="dv">25</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">25</span>],<span class="st">&quot;nbl_result_matrix_small.rds&quot;</span>)</span></code></pre></div>
<p>You’ll notice that there are no signs in this matrix (they’re just negative log p-values, which are always positive). We’ll have to assign signs by the correlation matrix next, then we will chunk the large matrix into smaller, flattened matrices that the shiny app can handle. For lower capacity machines/clusters, an alternative may be using the cor function.</p>
<p>In order to perfrom sign correction,fix the “Inf values” to a viewable value, and restore column and row names, postProcessLinRegMatrix() can be applied, yielding a final full matrix of the entire genome (Chr1-&gt;ChrX on both Axes). 300 has been used, although something a bit smaller will reduce saturation issues depending on the disparity between the lowest values in the matrix and 300. We’ll plot the result below, using complexheatmap and a custom designed function that takes large asymmetric distributions of values and pushes them into the [0,1] colorspace with white at 0.5, corresponding to zero, values between 0 and 0.5 corresponding to negative values, and values from 0.5 to 1 corresponding to positive values (signedRescale).</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>nbl_result_matrix_small&lt;-<span class="kw">readRDS</span>(<span class="st">&quot;nbl_result_matrix_small.rds&quot;</span>)</span>
<span id="cb7-2"><a href="#cb7-2"></a>nbl_result_matrix_small[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>##          [,1]       [,2]       [,3]       [,4]       [,5]
## [1,]      Inf   4.766959   3.743363   3.800939   3.800939
## [2,] 4.766959        Inf 199.883186 176.388907 176.388907
## [3,] 3.743363 199.883186        Inf 299.519384 299.519384
## [4,] 3.800939 176.388907 299.519384        Inf        Inf
## [5,] 3.800939 176.388907 299.519384        Inf        Inf</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>nbl_result_matrix_sign_corrected&lt;-<span class="kw">postProcessLinRegMatrix</span>(<span class="dt">input_matrix =</span> nbl_input_matrix[<span class="dv">1</span><span class="op">:</span><span class="dv">25</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">25</span>],<span class="dt">LM_mat =</span> nbl_result_matrix_small,<span class="dt">cor_type =</span> <span class="st">&quot;pearson&quot;</span>,<span class="dt">inf_replacement_val =</span> <span class="dv">300</span>)</span>
<span id="cb9-2"><a href="#cb9-2"></a>nbl_result_matrix_sign_corrected[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>##                      chr1_1_1000000 chr1_1000001_2000000 chr1_2000001_3000000
## chr1_1_1000000           300.000000             4.766959             3.743363
## chr1_1000001_2000000       4.766959           300.000000           199.883186
## chr1_2000001_3000000       3.743363           199.883186           300.000000
## chr1_3000001_4000000       3.800939           176.388907           299.519384
## chr1_4000001_5000000       3.800939           176.388907           299.519384
##                      chr1_3000001_4000000 chr1_4000001_5000000
## chr1_1_1000000                   3.800939             3.800939
## chr1_1000001_2000000           176.388907           176.388907
## chr1_2000001_3000000           299.519384           299.519384
## chr1_3000001_4000000           300.000000           300.000000
## chr1_4000001_5000000           300.000000           300.000000</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>nbl_result_matrix_sign_corrected[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>##                      chr1_1_1000000 chr1_1000001_2000000 chr1_2000001_3000000
## chr1_1_1000000           300.000000             4.766959             3.743363
## chr1_1000001_2000000       4.766959           300.000000           199.883186
## chr1_2000001_3000000       3.743363           199.883186           300.000000
## chr1_3000001_4000000       3.800939           176.388907           299.519384
## chr1_4000001_5000000       3.800939           176.388907           299.519384
##                      chr1_3000001_4000000 chr1_4000001_5000000
## chr1_1_1000000                   3.800939             3.800939
## chr1_1000001_2000000           176.388907           176.388907
## chr1_2000001_3000000           299.519384           299.519384
## chr1_3000001_4000000           300.000000           300.000000
## chr1_4000001_5000000           300.000000           300.000000</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>ComplexHeatmap<span class="op">::</span><span class="kw">Heatmap</span>(<span class="kw">signedRescale</span>(<span class="kw">as.matrix</span>(nbl_result_matrix_sign_corrected)),<span class="dt">col =</span> circlize<span class="op">::</span><span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="dv">0</span>,<span class="fl">0.5</span>,<span class="dv">1</span>),<span class="kw">c</span>(<span class="st">&quot;blue&quot;</span>,<span class="st">&quot;white&quot;</span>,<span class="st">&quot;red&quot;</span>)),<span class="dt">cluster_rows =</span> F,<span class="dt">cluster_columns =</span> F,<span class="dt">show_heatmap_legend =</span> F,<span class="dt">show_column_names =</span> F,<span class="dt">show_row_names =</span> F)</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAABklBMVEUAAP9pP/+vhv+4kv+5k/+8l/+/m//AnP/Eov/MrP/Mrf/PsP/RtP/Uuf/Xvf/l0///AAD/eln/gmH/h2f/iGn/iWn/jGz/j3D/lXf/lnj/mXv/m33/n4L/pIn/pYn/por/p4z/p43/qI3/qI7/qY//qo//q5H/rpX/r5X/sJf/sZn/spn/spr/s5r/s5v/tJz/tZ7/t6D/uKL/uqT/u6T/vKb/vKf/vaf/vaj/vqj/vqn/wKv/wKz/waz/wa3/wq3/wq7/w67/w6//xLD/xLH/xbH/xrL/xrP/x7T/x7X/yLX/yLb/ybb/ybf/yrf/yrj/y7n/y7r/zLr/zLv/zr3/zr7/z77/z7//0L//0MD/0cH/0cL/0sP/08T/1MX/1Mb/1cb/1sf/1sj/18n/2Mr/2cz/2s3/287/28//3M//3ND/3dH/3tL/3tP/39P/39T/4NT/4NX/4db/4df/4tf/4tj/49j/49n/5Nr/5Nv/5dv/5dz/5tz/5t3/597/6N//6OD/6eH/6uL//f3//v3///+lHKlaAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAF00lEQVR4nO3d2beOZRyH8V1knhqoSClFxlJ2pqioZEgUGYpEFEWGkG2m/u+Oer7XwU+3WreVZV3X4ePd7Pfj6Lfu4Rn50/6xkf/7F3jUE6iRQI0EaiRQI4Ea/Q00L91Lf7QaSa2PPpGeTE+lSWlKmpjGpQlpfJoxNDNNT9PS1ISnk5NAAgkkkEACCSSQQAIJ1AMIKrBakz4p+yLtHNqT9qbv04V0JV1N19L1shtpLN0aul12p9XdJJBAAgkkkEACCSSQQAL1AMLQBJV5rZakZUPvpPcSLH9Mv6Rf0/l0oexiwtPfq2r3uptJIIEEEkgggQQSSCCBBOoMhFGrCbQgLRxamlYnTGin0tlUq/xWdinhaSjGyuphDk8FEkgggQQSSCCBBBJIIIEEEkgggQQSSCCBBBJIoEcfCKdSsD6DVZ0FZbBaNYR1o83pq1Qv+5wtO1d2vuzy0JWyejEIT8EmkEACCSSQQAIJJJBAAvUAwiyWUyk7s2ts2cKyqKwK1fy0Mn2efkpnEr5/cxarK78/ZjHsJcMA9q93mAkkkEACCSSQQAIJJFAJhOUHnLpYWoapIiqYP15On6aT6ef0kM5qoNoKTx9s1BBIIIEEEkgggQQSSCCB/gsQbo3Cqe7VZVi1yNQFFVitSz8k7KWCFVY1wFYvcOAD2VXVXNWoA5tAAgkkkEACCSSQQAIJ1AMIb73ArbTYS7WnDLuisqqDqQsqsNqUdqf96Zt0pOy7dDSdGDpZhv+N0wlPsQglkEACCSSQQAIJJJBAAvUAwlv1sI6CQxWnyvCB7BrDr4SpCyqwejFhtxpmwA8TBr8t6bO0bwjYB9PhVseTQAIJJJBAAgkkkEACCdQDCG/txjpJ89A2PpAxBus3sMTUBRVY4eDLu2lj+iCBbWvKetXX6UD6NkEFT48lgQQSSCCBBBJIIIEEEqgH0KSEM+fYvVXf+QSrcs8XrDAeYeqCCqyWp7VpfdqQMOPtGNqV8B+TWW0fBPEUc5tAAgkkkEACCSSQQAIJ1ANoSsKJDlDU59cBlIc1IHaNYVUHUxdUYIWTMa+l1xOIsxaETW7b0pcJW+Pw9FASSCCBBBJIIIEEEkgggXoATUy4KBcU9f27OL+eh/gpjGU4ioJFG6zqYOqqT94/n7BwtChlQMMK0fsJW9Q+TniKuU0ggQQSSCCBBBJIIIEE6gE0LvUEwp8DCL8SvglWdTB1QQVWL6U3UxaA8C98lLAZDUdg8BQTmkACCSSQQAIJJJBAAgnUA2hCwts38H6PS2VY4QkQfgpAuFULp94xlmHXGFZ1MHVBBVavpqhg1KqP2G9PeIqLkAUSSCCBBBJIIIEEEkigHkDjE97515zFSiD8FJZ9cGsvfiUMQjiVgl1jWNXB1AUVWI0O4RZivIZxUyu88l0ggQQSSCCBBBJIIIEE6gE0I91KeP3f1bLLqXxpIMa2vBXkBE6qY/rJqfcdmJ8woeGbYFknKqOhmpOeTa+kNxKe4u8SSCCBBBJIIIEEEkigBwKamW4nfNWxMlyMGzX8FP4cb9vAZVS4CQq3RuFUN3ZYQQWrFpgqooL5Y27C2LI44Sn+LoEEEkgggQQSSCCBBBKoB9D0BCBMXU2gPGwC4U5ZvPUCZ7JxaxROdePUBYY1rFpk6oIKrGanF9Jz6e0kkEACCSSQQAIJJJBAAvUAmpbuJNx7e70MFvnoWGmFt/nV78/DYhDun8UAVp/qxmJQ1m8wX0EFVs+kWWlFEkgggQQSSCCBBBJIIIF6AE1NNdCNMgxrUcNPAeh0OlyGFSDcBIUFHqjgVDeAsmsMqzqYuqACq6fTW0kggQQSSCCBBBJIIIEE6gH0+M5iWNXB1AUVWOEtggIJJJBAAgkkkEACCSRQD6DJ6W662QpjWfkQwmfS8XQsYd8Z3tqNfWeY0HDcHrdG5dQ7TqVg1xhWdTB11W8pF0gggQQSSCCBBBJIIIF6ANl9EqiRQI0EaiRQI4EaCdRIoEZ/Ab7D+Qlre3RhAAAAAElFTkSuQmCC" /><!-- --> Finally, the whole genome matrix is too big to plot interactively without crashing most browsers using the plotly package. We’ll need to break things apart a bit. A final function will write chromosomal pair heatmaps to disk with genes from ensembl (hg19 coordinates) in encoded for each square in the matrix. Please only use this function on the WHOLE matrix, not on the small subset we have provided in documentation.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="cf">if</span>(<span class="op">!</span><span class="kw">dir.exists</span>(<span class="st">&quot;nbl_matrix_set&quot;</span>)){<span class="kw">dir.create</span>(<span class="st">&quot;nbl_matrix_set&quot;</span>)}</span>
<span id="cb14-2"><a href="#cb14-2"></a><span class="co">#setwd(&quot;nbl_matrix_set&quot;)</span></span>
<span id="cb14-3"><a href="#cb14-3"></a>doMC<span class="op">::</span><span class="kw">registerDoMC</span>()</span>
<span id="cb14-4"><a href="#cb14-4"></a><span class="co">#use ONLY the whole matrix with chromosomes 1-X, not the small subset provided for documentation purposes.</span></span>
<span id="cb14-5"><a href="#cb14-5"></a><span class="kw">createChromosomalMatrixSet</span>(<span class="dt">whole_genome_mat=</span>nbl_result_matrix_sign_corrected,<span class="dt">output_dir=</span><span class="st">&quot;nbl_matrix_set&quot;</span>,<span class="dt">prefix=</span><span class="st">&quot;nbl_&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a><span class="kw">list.files</span>(<span class="st">&quot;nbl_matrix_set&quot;</span>)</span></code></pre></div>
<p>There should be 529 of these particular files upon running the code. If there are not, don’t hesitate to run the code again. It can happen on a cluster. It is built to detect when chromosomal matrix is already written to disk.</p>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
